cheetah/accelerator.py

```diff
@@ -299,6 +307,354 @@ def defining_features(self) -> list[str]:
+                        (idx_s >= 0) & (idx_s < grid_shape[2])
+            
+            valid_indices = torch.stack([idx_x[valid_mask], idx_y[valid_mask], idx_s[valid_mask]], dim=0)
+            force_values = force_grid[tuple(valid_indices)]
+            return force_values, valid_mask
+
+        Fx_values, valid_mask_x = _get_force_values(grad_x)
+        Fy_values, valid_mask_y = _get_force_values(grad_y)
+        Fz_values, valid_mask_z = _get_force_values(grad_z)
+
+        # Compute interpolated forces
+        interpolated_forces = torch.zeros((particle_pos.shape[0], 3), device=grad_x.device)
+        values_x = cell_weights.view(-1)[valid_mask_x] * Fx_values * elementary_charge
+        values_y = cell_weights.view(-1)[valid_mask_y] * Fy_values * elementary_charge
+        values_z = cell_weights.view(-1)[valid_mask_z] * Fz_values * elementary_charge
+        
+        indices = torch.arange(particle_pos.shape[0]).repeat_interleave(8)
+        interpolated_forces.index_add_(0, indices[valid_mask_x], torch.stack([values_x, torch.zeros_like(values_x), torch.zeros_like(values_x)], dim=1))
+        interpolated_forces.index_add_(0, indices[valid_mask_y], torch.stack([torch.zeros_like(values_y), values_y, torch.zeros_like(values_y)], dim=1))
+        interpolated_forces.index_add_(0, indices[valid_mask_z], torch.stack([torch.zeros_like(values_z), torch.zeros_like(values_z), values_z], dim=1))
+
+        return interpolated_forces
+    
+
+    def track(self, incoming: ParticleBeam) -> ParticleBeam:
+        """
+        Track particles through the element. The input must be a `ParticleBeam`.
+
+        :param incoming: Beam of particles entering the element.
+        :return: Beam of particles exiting the element.
+        """
+        if incoming is Beam.empty:
+            return incoming
+        elif isinstance(incoming, ParticleBeam):
+            particles = self._cheetah_to_moments(incoming)
+            forces = self._read_forces(incoming)
+            dt = self._delta_t(incoming)
+            particles[:,1] += forces[:,0]*dt
+            particles[:,3] += forces[:,1]*dt
+            particles[:,5] += forces[:,2]*dt
+            particles = self._moments_to_cheetah(particles, incoming)
+            return ParticleBeam(
+                particles,
+                incoming.energy,
+                particle_charges=incoming.particle_charges,
+                device=particles.device,
+                dtype=particles.dtype,
+            )
+        else:
+            raise TypeError(f"Parameter incoming is of invalid type {type(incoming)}")
+
+
+    def split(self, resolution: torch.Tensor) -> list[Element]:
+    # TODO: Implement splitting for SpaceCharge properly, for now just returns the
+    # element itself
+        return [self]
+
+
+    @property
+    def is_skippable(self) -> bool:
+        return False
+
+    def plot(self, ax: matplotlib.axes.Axes, s: float) -> None:
```

Comment:
I need to understand better the plot functions of cheetah to implement something here. The only thing I know is there will be an issue given the fact that spacechargekick elements add 'artificial' lenght to the beam. @cr-xu @jank324

Reply:
@greglenerd So I think we need to be careful with the `length` property. In all existing elements, `length` represents the physical length of the beam pipe and Cheetah actually relies on this assumption in some places. I would suggest that `SpaceChargeKick.length = 0.0` all the time (because it doesn't add length to the beam pipe), and that the property of `SpaceChargeKick` currently named `length` is renamed to something like `effect_length`. Does that make sense?

Based on this, I think `SpaceChargeKick` could be plotted with something like `plt.axvspan` as a background highlight from `s - effect_length` to `s`.